### ISSUE TITLE
jOOQ generates invalid SQL when using correlated sub queries that use row values on Oracle XE 18

### ISSUE THREAD

**Issue by alf** [jOOQ/jOOQ#17038]
### Expected behavior

Generated SQL should work.

### Actual behavior

Generated SQL causes ORA-22922: nonexistent LOB value.

### Steps to reproduce the problem

See [JavaTest.java](https://github.com/alf/jOOQ-mcve/blob/92c6233f2fb01116bcaf059cacd2996c7f472347/jOOQ-mcve-java-oracle/src/test/java/org/jooq/mcve/test/java/oracle/JavaTest.java) for my MCVE.

In short, the following queries does not work against Oracle 18 XE. Note that they do work on Oracle 21 and 23.

```java
ctx.select(
    select(row(val(1), val(1))).asField())
.from(xmltable("1 to 11").column("column_value", INTEGER))
.fetch();
```

Result from logs:

```sql
org.jooq.exception.DataAccessException: SQL [select (select json_array(nvl(null, ?), nvl(null, ?) null on null returning clob) "nested" from DUAL) from xmltable(? columns "column_value" number(10))]; ORA-22922: nonexistent LOB value
```

```java
ctx.select(
    select(row(val(1), val(1))).asMultiset())
.from(xmltable("1 to 11").column("column_value", INTEGER))
.fetch();
``` 

Result from logs:

```sql
org.jooq.exception.DataAccessException: SQL [select (select coalesce(json_arrayagg(json_array(t."v0" format json null on null returning clob) format json returning clob), json_array(returning clob)) from (select json_array(nvl(null, ?), nvl(null, ?) null on null returning clob) "v0" from DUAL) t having count(*) = count(*)) from xmltable(? columns "column_value" number(10))]; ORA-22922: nonexistent LOB value
```

### jOOQ Version

3.19.10

### Database product and version

Oracle 18 XE

### Java Version

OpenJDK 17

**Comment by lukaseder** [comment]
Thanks a lot for your report. I'll look into  this soon.

**Comment by lukaseder** [comment]
I can reproduce this. Will look into it right away.

**Comment by lukaseder** [comment]
There's a fun bug in Oracle 18c, somehow related to your usage of `xmltable` along with jOOQ's `RETURNING CLOB` clause. You can see it here:
https://dbfiddle.uk/RSE3ioVQ

`RETURNING CLOB` seems to truncate the result set in some way:

![image](https://github.com/user-attachments/assets/d1793fed-2eec-4060-b9fc-4bb823776cd2)

It doesn't happen with `CONNECT BY LEVEL <= 10` or other ways to generate rows. And it's not related to this issue here, nor does it reproduce with 23ai, so I'll just ignore it 😅

**Comment by lukaseder** [comment]
Welp, it does seem to be the same problem, though. As soon as I do this, the problem goes away:

```sql
select 
  rownum, 
  (select json_array(1, 1 null on null returning clob) "nested" from DUAL) || to_clob('')
from xmltable('1 to 11' columns "column_value" number(10))
```

Observe the dummy concatenation `|| to_clob('')`

See also:
https://dbfiddle.uk/ua3FQNA0

This is an exceptionally ugly workaround for an Oracle bug, probably the worst I've ever thought of... I'll check if I find any other way, before adding this.

**Comment by lukaseder** [comment]
Also, it doesn't really work. jOOQ can only render this with reasonable efforts:

```sql
select 
  rownum, 
  (select json_array(1, 1 null on null returning clob) || to_clob('') "nested" from DUAL)
from xmltable('1 to 11' columns "column_value" number(10))
```

But that doesn't fix the issue. Only the concatenation at the top level does

**Comment by lukaseder** [comment]
I'm convinced this is some scalar subquery caching bug. The problem doesn't appear like this:

```sql
SELECT x
FROM dual, (SELECT to_clob('123412341234') x FROM dual) x
CONNECT BY LEVEL <= 11;
```

The scalar subquery cache has always been a bit of a controversial feature:
https://blog.jooq.org/oracle-scalar-subquery-caching/

But I'm surprised by the severity of this bug! The feature has been added in 10g I believe? But people must have projected `CLOB` values from scalar subqueries, I believe, so this must have been an issue many times?

**Comment by lukaseder** [comment]
OK, no that hint was wrong. It's more related to how ojdbc interacts with the cursor. This works:

```java
try (ResultSet rs = connection.createStatement().executeQuery("""
    SELECT (SELECT to_clob('123412341234') x FROM dual)
    FROM dual
    CONNECT BY LEVEL <= 11
    """)) {
    while (rs.next())
        System.out.println(rs.getClob(1).getSubString(1, (int) rs.getClob(1).length()));
}
```

But this exposes the problem:

```java
try (ResultSet rs = connection.createStatement().executeQuery("""
    SELECT (SELECT to_clob('123412341234') x FROM dual)
    FROM dual
    CONNECT BY LEVEL <= 11
    """)) {
    while (rs.next())
        System.out.println(rs.getString(1));
}
```

**Comment by lukaseder** [comment]
Oh wow, the bug seems to be that if I call `rs.getString(1)`, or `clob.free()` in the loop, then the problem appears. If I don't free the `Clob` values, or delay the freeing, then the problem disappears.

Well, I think we can work with that.

**Comment by alf** [comment]
Note that the usage of `xmltable` is there just because I couldn’t be bothered setting up tables with data for the test case. We don’t have that in our code anywhere.

**Comment by lukaseder** [comment]
Yeah sure, I figured eventually

**Comment by alf** [comment]
Really looking forward to upgrading our Oracle, but thanks to jOOQ it’s not that bad. Thank you for your hard work.

**Comment by lukaseder** [comment]
> Really looking forward to upgrading our Oracle

Yeah, 23ai did bring a lot of improvements and fixed a ton of JSON related bugs. If you ever find something, you can report it here as well:
https://forums.oracle.com/ords/apexds/domain/dev-community/category/oracle-database-free

They've started this channel as a means of helping power users get heard more quickly, rather than going through the usually not-so-efficient commercial support channels. At least, it works well this way for me.

Continuing this journey here, I've tried SQL*Plus, and the problem persists there:

```
SQL> SELECT (SELECT to_clob('123412341234') x FROM dual)
  2  FROM dual
  3  CONNECT BY LEVEL <= 11;

(SELECTTO_CLOB('123412341234')XFROMDUAL)
--------------------------------------------------------------------------------
123412341234
ERROR:
ORA-22922: nonexistent LOB value
```

So, I guess it isn't a problem of ojdbc...

**Comment by lukaseder** [comment]
Well, one thing I had noticed was that

1. You seem to have deliberately chosen to fetch 11 rows
2. The default fetch size in ojdbc is 10

So, when I tried to increase it to `1000000` or so, the problem goes away (for result sets smaller than `1000000`, of course). This means that ojdbc will use much more memory, but perhaps, you can use this as a workaround for now?

**Comment by lukaseder** [comment]
I've documented my findings here as well:
https://stackoverflow.com/q/78900296/521799

Latest update, it's clearly related to the fetch size. Delaying the freeing of clobs by exactly the fetch size seems to work!

```java
try (Statement s = connection.createStatement()) {
    try (ResultSet rs = s.executeQuery(
        """
        SELECT (SELECT to_clob('123412341234') x FROM dual)
        FROM dual
        CONNECT BY LEVEL <= 100
        """
    )) {
        Deque<Clob> clobs = new ArrayDeque<>();

        while (rs.next()) {
            Clob clob = rs.getClob(1);
            clobs.add(clob);
            System.out.println(clob.getSubString(1, (int) clob.length()));

            int size = clobs.size() - s.getFetchSize();
            while (size --> 0)
                clobs.pollFirst().free();
        }

        for (Clob clob : clobs)
            clob.free();
    }
}
```

**Comment by lukaseder** [comment]
And the report to Oracle's community managers:
https://forums.oracle.com/ords/apexds/post/ora-22922-nonexistent-lob-value-in-oracle-18c-when-fetching-7625

**Comment by lukaseder** [comment]
Fun fact, Oracle 18c is no longer supported by Oracle:
https://en.wikipedia.org/wiki/Oracle_Database

I didn't realise this, thought it was still supported.

I guess this depends on what you're expecting. The workarounds seem reasonable in general, especially the fact that the fetch size can be changed on a per-query basis in jOOQ.

I don't really want to mess around with `Clob` types for *all* Oracle users and *all* JSON usage. This could be a thing that users can opt in to, e.g. a `Settings` that specifies how to read `JSON` and `CLOB` data to override the current `ResultSet.getString()` calls.

What do you think?

**Comment by alf** [comment]
Our test and prod-environment is running Oracle 19c, but since we don’t have a docker image for this (licensing problems) our development is leveraging 18 XE.

We could switch to 23 for development but it feels iffy to use a newer version than what we have running in production.

I believe (hope) we’re upgrading our production environments this fall, which would fix our problems here.

From my understanding, the workaround should suffice. Just to be sure I understand: We need to set the fetch size to the max value of our page size (+1)?

**Comment by lukaseder** [comment]
> From my understanding, the workaround should suffice. Just to be sure I understand: We need to set the fetch size to the max value of our page size (+1)?

Yes, that's my understanding. If your paginating, then this workaround should work very well.
